packages/compiler/server/serverlib.ts

```diff
@@ -403,6 +408,42 @@ export function createServer(host: ServerHost): Server {
     }
   }
 
+  async function getFoldingRanges(params: FoldingRangeParams): Promise<FoldingRange[]> {
+    const file = await compilerHost.readFile(getPath(params.textDocument));
+    const ast = parse(file);
+    const ranges: FoldingRange[] = [];
+    getRangesForNode(ast);
+
+    function getRangesForNode(node: Node) {
+      if ("decorators" in node && node.decorators.length > 0) {
+        const startPos = node.decorators[0].pos;
+        const endPos = node.decorators[node.decorators.length - 1].end;
+        const nodeEndPos = node.end;
+        let startLine = file.getLineAndCharacterOfPosition(startPos).line;
+        let endLine = file.getLineAndCharacterOfPosition(endPos).line;
+        if (startLine !== endLine) {
+          ranges.push({ startLine, endLine });
+        }
+        startLine = file.getLineAndCharacterOfPosition(skipTrivia(file.text, endPos)).line;
```

Comment:
There's a fair amount of duplication here. Consider:

- factoring out a helper to add a range from (startPos: number, endPos: number).
- adjusting a startPos when you add the decorator range and then reuse the same code to add the main node range as the case without decorators.

Reply:
sounds good

Reply:
We can still eliminate more repetition and make it easier to follow.

This is what I meant by factoring out a helper, nest this at the bottom getFoldingRanges, and then there is only one place where we need to convert between positions and line+char.

``` js
function addRange(startPos: number, endPos: number) {
  const start = file.getLineAndCharacter(pos);
  const end = file.getLineAndCharacter(end);
  if (start.line !=== end.line) { // don't add ranges that only span one line.
    ranges.push({
      startLine: start.line,
      startCharacter: star.character,
      endLine: end.line,
      endCharacter: end.character
    });
  }
}
```

Now we only have to deal with position to line+character in and not adding single-line ranges in one place and we can write something like this:

``` js
function getRangesForNode(node: Node) {
  // If a node has decorators fold them separately.
  let nodeStart = node.pos;
  if ("decorators" in node && node.decorators.length > 0) { 
     const decoratorEnd = node.decorators[node.decorators.length - 1].end
     addRange(startPos, decoratorEnd);
     nodeStart = skipTrivia(decoratorEnd).
  }

  addRange(nodeStart, node.end);
  visitChildren(node, getRangesForNode);
}
```

I purposely removed the check for CadlScript because it occurs to me that this can be handled once by visiting the children of the root, but not the root:

```diff
< getRangesForNode(ast);
> visitChildren(ast, getRangesForNode);
```

And finally a nit: a better name for getRangesForNode would be addRangesForNode since it adds them to a list and doesn't return them.